### PR TITLE
testEqualDerivation: init

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation for this change

The `environment.noXlibs` overlay got me thinking that if we want to take that seriously (ie #29135), we'll want to control rebuilds. For example, we'll want to prevent regressions of, say, #126611.

`testEqualDerivation` makes this easy to check and reports the differences nicely with `nix-diff`.

See for yourself:

```
gh pr checkout 126718
nix-build --expr 'with import <nixpkgs> {}; testEqualDerivation "ghc does not have to be rebuilt when environment.noXlibs is set." ghc (nixos { environment.noXlibs = true;})._module.args.pkgs.ghc'
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/hello/default.nix [hunk 1]
```diff
@@ -16,8 +18,16 @@ stdenv.mkDerivation rec {
 
   doCheck = true;
 
-  passthru.tests.version =
-    testVersion { package = hello; };
+  passthru.tests = {
+    version =
+      testVersion { package = hello; };
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/hello/default.nix:23]:
```suggestion
    version = testVersion { package = hello; };
```

File: pkgs/build-support/test-equal-derivation.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    {
+      inherit assertion drvA drvB;
+      nativeBuildInputs = [ nix-diff ];
+    } ''
+      echo "$assertion"
+      echo "However, the derivations differ:"
+      echo
+      echo nix-diff $drvA $drvB
+      nix-diff $drvA $drvB
+      false
```
Comment (SuperSandro2000) [inline pkgs/build-support/test-equal-derivation.nix:30]:
```suggestion
      exit
```

File: pkgs/build-support/test-equal-derivation.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  This can be used to make sure that a certain difference of configuration,
+  such as the presence of an overlay does not cause a cache miss.
+*/
+assertion: a: b:
+let
+  drvA = builtins.unsafeDiscardOutputDependency a.drvPath or (throw "testEqualDerivation second argument must be a package");
+  drvB = builtins.unsafeDiscardOutputDependency b.drvPath or (throw "testEqualDerivation third argument must be a package");
+  name =
+    if a?name
```
Comment (SuperSandro2000) [inline pkgs/build-support/test-equal-derivation.nix:13]:
```suggestion
    if a ? name
```

File: pkgs/applications/misc/hello/default.nix [hunk 4]
```diff
@@ -16,8 +18,16 @@ stdenv.mkDerivation rec {
...
+
+    invariant-under-noXlibs =
+      testEqualDerivation
+        "hello must not be rebuilt when environment.noXlibs is set."
+        hello
+        (nixos { environment.noXlibs = true; })._module.args.pkgs.hello;
```
Comment (raboof) [inline pkgs/applications/misc/hello/default.nix:29]: Are we 'allowed' to refer to `_module`?

**Review by raboof — APPROVED** [review]
I won't pretend to understand this fully, but it looks useful and harmless.

File: pkgs/build-support/test-equal-derivation.nix (hunk 2 above)
Comment (roberth) [reply pkgs/build-support/test-equal-derivation.nix:30]: It needs to fail. I'll do `exit 1`.

File: pkgs/applications/misc/hello/default.nix (hunk 4 above)
Comment (roberth) [reply pkgs/applications/misc/hello/default.nix:29]: Preferably not. `nixos` exposes `pkgs` directly, so I'll use that. (`evalModules` doesn't)

Comment (roberth) [reply pkgs/applications/misc/hello/default.nix]: Done. Thanks!

**Comment by roberth** [comment]
Rebased and improved doc a bit. Will merge if ofborg agrees.

**Pull request merged**